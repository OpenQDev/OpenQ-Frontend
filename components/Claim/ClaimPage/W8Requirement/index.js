import React, { useState, useContext, useEffect } from 'react';
import Link from 'next/link';
import { CheckIcon, UploadIcon } from '@primer/octicons-react';
import W8FormModal from './W8FormModal';
import axios from 'axios';
import {
  EMAIL_NOT_SENT,
  INVALID_EMAIL_CLIENT,
  INVALID_EMAIL_FREELANCER,
} from '../../../../constants/invoiceableResponses';
import StoreContext from '../../../../store/Store/StoreContext';
import { getW8Approved } from '../../../../services/utils/lib';
import LoadingIcon from '../../../Loading/ButtonLoadingIcon';
import FreelancerDetails from '../../../User/InvoicingDetailsTab/FreelancerDetails';
const W8Requirement = ({ bounty }) => {
  const [loading, setLoading] = useState(false);
  const [appState] = useContext(StoreContext);
  const [file, setFile] = useState(null);
  const [w8formResponse, setW8FormResponse] = useState('');
  const [showModal, setShowModal] = useState(false);
  const { accountData } = appState;
  const pending = bounty.requests?.nodes.some((node) => node.requestingUser.id === accountData.id);
  const [sent, setSent] = useState(pending);
  const profileLink = `${process.env.NEXT_PUBLIC_BASE_URL}/user/${accountData.id}?tab=Invoicing (Freelancer)`;
  const [w8Approved, setW8Approved] = useState(false);
  const [currentRequest, setCurrentRequest] = useState(null);

  const noEmail = !accountData?.invoicingEmail;
  useEffect(() => {
    const W8Approved = getW8Approved(bounty, accountData);
    setW8Approved(W8Approved);

    const getPrivateRequest = async () => {
      const request = bounty.requests?.nodes?.find((node) => node.requestingUser.id === accountData.id);

      if (request) {
        try {
          const privateRequest = await appState.openQPrismaClient.getPrivateRequest(request.id);
          setCurrentRequest(privateRequest?.message);
        } catch (e) {
          appState.logger.error(e);
        }
      }
    };
    getPrivateRequest();
  }, [bounty, accountData]);

  const handleFileChange = (e) => {
    setFile(e.target.files[0]);
    setSent(false);
    setW8FormResponse('');
  };

  const handleResult = (result) => {
    if (result?.data?.message) {
      setW8FormResponse(result.data.message);
    } else setW8FormResponse(EMAIL_NOT_SENT);

    setLoading(false);
  };
  const handleSend = async (e) => {
    e.preventDefault();
    setLoading(true);
    setW8FormResponse('LOADING');
    const formData = new FormData();
    setSent(true);
    formData.append('file', file);
    setFile(null);
    try {
      const result = await axios.post(
        `${process.env.NEXT_PUBLIC_INVOICE_URL}/taxform?id=${bounty.bountyAddress}`,
        formData,
        {
          withCredentials: true,
          headers: {
            'content-type': 'multipart/form-data',
          },
        }
      );
      handleResult(result);
    } catch (e) {
      setLoading(false);
      setW8FormResponse(EMAIL_NOT_SENT);
      appState.logger.error('w8form.js1', e);
    }
    e.target.reset();
  };

  const w8formResponseOptions = {
    '': {
      MessageHTML: () => <></>,
    },
    [INVALID_EMAIL_CLIENT]: {
      MessageHTML: () => (
        <>
          The organization's invoicing email is invalid.
          {w8formResponse?.invoicingEmail && `The email entered was  ${w8formResponse.invoicingEmail}.`} Please contact
          them or ask for help in our{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
          .
        </>
      ),
    },
    [INVALID_EMAIL_FREELANCER]: {
      MessageHTML: () => (
        <>
          Your email is an invalid format. Please head to your{' '}
          <Link className='underline' href={profileLink}>
            profile
          </Link>{' '}
          and correct it or ask for help in our{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
          .
        </>
      ),
    },
    EMAIL_SENT: {
      successW8Form: true,
      MessageHTML: () => (
        <>
          We have sent your tax document to the organization. Please wait until they have viewed and confirmed them. If
          they request changes, you will be notified via your contact email. If you have any questions, feel free to
          reach out in our{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
          .
        </>
      ),
    },
    NO_DEPOSITS: {
      MessageHTML: () =>
        "Funder hasn't made any deposits yet, please wait for money to be deposited before sending W8/W9 form.",
    },
    'no deposits': {
      MessageHTML: () =>
        "Funder hasn't made any deposits yet, please wait for money to be deposited before sending W8/W9 form.",
    },

    NOT_WINNER: {
      MessageHTML: () => (
        <>
          You are not the winner of this bounty
          <br />
          <br />
          If you believe this is an error, please reach out to us via{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
        </>
      ),
    },

    SECURITY_ERROR: {
      MessageHTML: () => (
        <>
          Your pdf failed our security scan, please reach out to us via{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
        </>
      ),
    },
    FILE_TOO_LARGE: {
      MessageHTML: () => (
        <>
          Your pdf was too large, must be under 10 mb threshold. Please reach out to us via{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
          .
        </>
      ),
    },

    NOT_PDF: {
      MessageHTML: () => <>Your file is not a pdf.</>,
    },
    EMAIL_NOT_SENT: {
      MessageHTML: () => (
        <>
          We are very sorry but it seems we are having difficulties with our email server, please try again or ask for
          support on our{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
          .
        </>
      ),
    },
    EMAIL_LIMIT_REACHED: {
      MessageHTML: () => (
        <>
          You have reached the limit of 4 emails per day, please try again tomorrow or ask for support on our{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
          .
        </>
      ),
    },
    LOADING: {
      MessageHTML: () => <>Scanning and sending your tax form to your client.</>,
    },
  };

  const MessageHTML = w8formResponseOptions[w8formResponse]?.MessageHTML || (() => <></>);
  return (
    <section className='flex flex-col gap-3'>
      <h4 className='text-2xl flex content-center items-center gap-2 border-b border-gray-700 pb-2'>
        Form W8/W9*
        <div
          className={`${
            w8Approved ? 'border-green bg-green-inside' : 'bg-info border-2 border-info-strong'
          } text-sm px-2 border rounded-full h-6`}
        >
          {w8Approved ? 'Approved' : 'Required'}
        </div>
      </h4>
      {w8Approved ? (
        <div className='border-green bg-green-inside border p-4 rounded-sm'> Your tax form was accepted</div>
      ) : (
        <>
          <div>
            {!w8Approved && currentRequest && (
              <div className='bg-info border-info-strong rounded-sm border p-4 my-4'>
                Your tax form was not accepted. {currentRequest}
              </div>
            )}
            <div>
              Please complete and upload a W-8/W-9 form. Choose one of five types, depending on your entity. We
              encourage you to consult with you own tax or financial adviser to determine which form is appropriate for
              you or ask in our
              <div>
                <Link
                  href={'https://discord.gg/puQVqEvVXn'}
                  rel='noopener norefferer'
                  target='_blank'
                  className='text-link-colour hover:underline col-span-2'
                >
                  discord
                </Link>{' '}
                for help.
              </div>
            </div>
            {noEmail && (
              <>
                <p>Please add an email to your profile so that we can send you a copy of the your submitted form.</p>
                <FreelancerDetails slim={true} emailOnly={true} />
              </>
            )}
          </div>
          <div>
            Explore our W8/W9 templates{' '}
            <button className='text-blue-500 hover:underline' onClick={() => setShowModal(true)}>
              here
            </button>
            .
          </div>
          <div className='font-semibold flex gap-2 group w-fit'>Upload</div>
          <form onSubmit={handleSend} className='flex gap-2 items-center flex-wrap md:flex-nowrap'>
            {sent && (
              <div
                className={`relative flex w-fit whitespace-nowrap gap-2 z-20 h-8 items-center justify-center text-center btn-default-disabled cursor-not-allowed`}
              >
                Pending
              </div>
            )}
            <label htmlFor='file input' className={`relative flex items-center`}>
              <div
                className={`relative flex w-fit whitespace-nowrap gap-2 z-20 h-8 items-center justify-center text-center ${
                  noEmail ? 'cursor-not-allowed' : 'cursor-pointer'
                } ${file || noEmail ? 'btn-primary' : 'btn-requirements'}`}
              >
                {file ? (
                  <>
                    <CheckIcon size={16} />
                    Change File
                  </>
                ) : sent ? (
                  <>
                    <UploadIcon size={16} />
                    Upload Updated W8/W9
                  </>
                ) : (
                  <>
                    <UploadIcon size={16} />
                    Choose File
                  </>
                )}
              </div>
              <input
                onChange={handleFileChange}
                disabled={loading || noEmail}
                type='file'
                className='absolute invisible w-full top-0 bottom-0 z-10'
                id='file input'
              />
            </label>
            <div className='border border-web-gray w-full flex items-center font-semibold h-8 px-2 rounded-sm'>
              {file?.name}
            </div>
            <button
              disabled={!file}
              className={`flex gap-2 h-8 items-center justify-center text-center
                ${file ? 'btn-requirements cursor-pointer' : 'btn-default cursor-not-allowed'}
              `}
            >
              {sent ? (loading ? 'Sending' : 'Sent') : 'Send'}
              {loading && <LoadingIcon />}
            </button>
          </form>
          <div className=''>
            *W-8 forms are{' '}
            <Link
              href={'https://www.irs.gov/'}
              rel='noopener norefferer'
              target='_blank'
              className='text-link-colour hover:underline col-span-2'
            >
              Internal Revenue Service
            </Link>{' '}
            (IRS) forms that foreign individuals and businesses must file to verify their country of residence for tax
            purposes, certifying that they qualify for a lower rate of tax withholding.
          </div>
        </>
      )}
      {w8formResponse && (
        <div className='bg-info border-info-strong border p-4 rounded-sm'>
          <MessageHTML />
        </div>
      )}
      {showModal && <W8FormModal setShowModal={setShowModal} />}
    </section>
  );
};

export default W8Requirement;
