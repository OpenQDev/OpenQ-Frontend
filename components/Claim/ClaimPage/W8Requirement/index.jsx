import React, { useState, useContext, useEffect } from 'react';
import Link from 'next/link';
import { CheckIcon, UploadIcon } from '@primer/octicons-react';
import W8FormModal from './W8FormModal';
import axios from 'axios';
import {
  EMAIL_NOT_SENT,
  INVALID_EMAIL_CLIENT,
  INVALID_EMAIL_FREELANCER,
} from '../../../../constants/invoiceableResponses';
import StoreContext from '../../../../store/Store/StoreContext';
import { getW8Approved } from '../../../../services/utils/lib';
import LoadingIcon from '../../../Loading/ButtonLoadingIcon';
import FreelancerDetails from '../../../User/InvoicingDetailsTab/FreelancerDetails';
import ModalDefault from '../../../Utils/ModalDefault';
const W8Requirement = ({ bounty }) => {
  const [loading, setLoading] = useState(false);
  const [appState] = useContext(StoreContext);
  const [file, setFile] = useState(null);
  const [w8formResponse, setW8FormResponse] = useState('');
  const [showModal, setShowModal] = useState(false);
  const { accountData } = appState;
  const pending = bounty.requests?.nodes.some((node) => node.requestingUser.id === accountData.id);
  const [sent, setSent] = useState(pending);
  const profileLink = `${process.env.NEXT_PUBLIC_BASE_URL}/user/${accountData.id}?tab=Invoicing (Freelancer)`;
  const [w8Approved, setW8Approved] = useState(false);
  const [showTaxModal, setShowTaxModal] = useState(false);
  const [currentRequest, setCurrentRequest] = useState(null);
  const [compliance, setCompliance] = useState({ signed: null, dated: null, relevant: null });

  const noEmail = !accountData?.invoicingEmail;
  useEffect(() => {
    const W8Approved = getW8Approved(bounty, accountData);
    setW8Approved(W8Approved);

    const getPrivateRequest = async () => {
      const request = bounty.requests?.nodes?.find((node) => node.requestingUser.id === accountData.id);

      if (request) {
        try {
          const privateRequest = await appState.openQPrismaClient.getPrivateRequest(request.id);
          setCurrentRequest(privateRequest?.message);
        } catch (e) {
          appState.logger.error(e);
        }
      }
    };
    getPrivateRequest();
  }, [bounty, accountData]);

  const handleFileChange = (e) => {
    setFile(e.target.files[0]);
    setSent(false);
    setW8FormResponse('');
  };

  const handleResult = (result) => {
    if (result?.data?.message) {
      setW8FormResponse(result.data.message);
    } else setW8FormResponse(EMAIL_NOT_SENT);

    setLoading(false);
  };
  const handleSend = async (e) => {
    e.preventDefault();
    setLoading(true);
    setW8FormResponse('LOADING');
    const formData = new FormData();
    setSent(true);
    formData.append('file', file);
    setFile(null);
    setCompliance({});
    setShowTaxModal();
    try {
      const result = await axios.post(
        `${process.env.NEXT_PUBLIC_INVOICE_URL}/taxform?id=${bounty.bountyAddress}`,
        formData,
        {
          withCredentials: true,
          headers: {
            'content-type': 'multipart/form-data',
          },
        }
      );
      handleResult(result);
    } catch (e) {
      setLoading(false);
      setW8FormResponse(EMAIL_NOT_SENT);
      appState.logger.error('w8form.js1', e);
    }
    e.target.reset();
  };

  const w8formResponseOptions = {
    '': {
      MessageHTML: () => <></>,
    },
    [INVALID_EMAIL_CLIENT]: {
      MessageHTML: () => (
        <>
          The organization's invoicing email is invalid.
          {w8formResponse?.invoicingEmail && `The email entered was  ${w8formResponse.invoicingEmail}.`} Please contact
          them or ask for help in our{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
          .
        </>
      ),
    },
    [INVALID_EMAIL_FREELANCER]: {
      MessageHTML: () => (
        <>
          Your email is an invalid format. Please head to your{' '}
          <Link className='underline' href={profileLink}>
            profile
          </Link>{' '}
          and correct it or ask for help in our{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
          .
        </>
      ),
    },
    EMAIL_SENT: {
      successW8Form: true,
      MessageHTML: () => (
        <>
          We have sent your tax document to the organization. Please wait until they have viewed and confirmed them. If
          they request changes, you will be notified via your contact email. If you have any questions, feel free to
          reach out in our{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            Discord
          </a>
          .
        </>
      ),
    },
    NO_DEPOSITS: {
      MessageHTML: () =>
        "Funder hasn't made any deposits yet, please wait for money to be deposited before sending W8/W9 form.",
    },
    'no deposits': {
      MessageHTML: () =>
        "Funder hasn't made any deposits yet, please wait for money to be deposited before sending W8/W9 form.",
    },

    NOT_WINNER: {
      MessageHTML: () => (
        <>
          You are not the winner of this bounty
          <br />
          <br />
          If you believe this is an error, please reach out to us via{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
        </>
      ),
    },

    SECURITY_ERROR: {
      MessageHTML: () => (
        <>
          Your pdf failed our security scan, please reach out to us via{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
        </>
      ),
    },
    FILE_TOO_LARGE: {
      MessageHTML: () => (
        <>
          Your pdf was too large, must be under 10 mb threshold. Please reach out to us via{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
          .
        </>
      ),
    },

    NOT_PDF: {
      MessageHTML: () => <>Your file is not a pdf.</>,
    },
    EMAIL_NOT_SENT: {
      MessageHTML: () => (
        <>
          We are very sorry but it seems we are having difficulties with our email server, please try again or ask for
          support on our{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
          .
        </>
      ),
    },
    EMAIL_LIMIT_REACHED: {
      MessageHTML: () => (
        <>
          You have reached the limit of 4 emails per day, please try again tomorrow or ask for support on our{' '}
          <a target={'_blank'} className='underline' href='https://discord.gg/puQVqEvVXn' rel='noreferrer'>
            discord
          </a>
          .
        </>
      ),
    },
    LOADING: {
      MessageHTML: () => <>Scanning and sending your tax form to your client.</>,
    },
  };
  const isCompliant = compliance.dated && compliance.signed && compliance.relevant;
  const submitBtn = (
    <button
      disabled={!isCompliant}
      className={`flex gap-2 h-8 items-center justify-center text-center
    ${isCompliant ? 'btn-requirements cursor-pointer' : 'btn-default cursor-not-allowed'}
  `}
    >
      {sent ? (loading ? 'Sending' : 'Sent') : 'Send'}
      {loading && <LoadingIcon />}
    </button>
  );
  const updateCompliance = (e) => {
    const id = e.target.id;
    const value = e.target.checked;
    setCompliance((prev) => ({ ...prev, [id]: value }));
  };
  const openModal = (e) => {
    e.preventDefault();
    setShowTaxModal(true);
  };

  const MessageHTML = w8formResponseOptions[w8formResponse]?.MessageHTML || (() => <></>);
  return (
    <section className='flex flex-col gap-3'>
      <h4 className='text-2xl flex content-center items-center gap-2 border-b border-gray-700 pb-2'>
        Form W8/W9*
        <div
          className={`${
            w8Approved ? 'border-green bg-green-inside' : 'bg-info border-2 border-info-strong'
          } text-sm px-2 border rounded-full h-6`}
        >
          {w8Approved ? 'Approved' : 'Required'}
        </div>
      </h4>
      {w8Approved ? (
        <div className='border-green bg-green-inside border p-4 rounded-sm'> Your tax form was accepted</div>
      ) : (
        <>
          <div>
            {sent && (
              <div className='bg-blue-200 text-blue-900 border-blue-500 border p-4 px-4 my-2 rounded-sm'>
                Thank you for submitting your tax form. Kindly note that OpenQ is not involved in the reviewing process,
                which may take up to 19 days. We appreciate your patience and will notify you as soon as the process is
                complete.
              </div>
            )}
            {!w8Approved && currentRequest && (
              <div className='bg-info border-info-strong rounded-sm border p-4 my-4'>
                The organization has requested changes on the tax form you submitted. Please review the requested change
                below, update your tax forms, and re-upload them here.
                <br />
                <br />
                <br />
                <div className='flex gap-2 font-bold'>
                  Requested change: <div className='text-gray-400 font-normal'>{currentRequest}</div>
                </div>
              </div>
            )}
            <div>
              Please complete and upload a W-8/W-9 form (must be in PDF format). Choose one of five types, depending on
              your entity. We encourage you to consult with you own tax or financial adviser to determine which form is
              appropriate for you or ask in our
              <div>
                <Link
                  href={'https://discord.gg/puQVqEvVXn'}
                  rel='noopener norefferer'
                  target='_blank'
                  className='text-link-colour hover:underline col-span-2'
                >
                  discord
                </Link>{' '}
                for help.
              </div>
            </div>
            {noEmail && (
              <>
                <p>Please add an email to your profile so that we can send you a copy of the your submitted form.</p>
                <FreelancerDetails slim={true} emailOnly={true} />
              </>
            )}
          </div>
          <div>
            Explore our W8/W9 templates{' '}
            <button className='text-blue-500 hover:underline' onClick={() => setShowModal(true)}>
              here
            </button>
            .
          </div>
          <div>✅ PDF File must be less than 6 MB</div>
          <div>🔔 Don't forget to add your signature and date to your W8 or W9 Form!</div>
          <form onSubmit={handleSend} className='flex gap-2 items-center flex-wrap md:flex-nowrap'>
            {sent && (
              <div
                className={`relative flex w-fit whitespace-nowrap gap-2 h-8 items-center justify-center text-center btn-default-disabled cursor-not-allowed`}
              >
                Under Review
              </div>
            )}
            <label htmlFor='file input' className={`relative flex items-center`}>
              <div
                className={`relative flex w-fit whitespace-nowrap gap-2 h-8 items-center justify-center text-center ${
                  noEmail ? 'cursor-not-allowed' : 'cursor-pointer'
                } ${file || noEmail ? 'btn-primary' : 'btn-requirements'}`}
              >
                {file ? (
                  <>
                    <CheckIcon size={16} />
                    Change PDF File
                  </>
                ) : sent ? (
                  <>
                    <UploadIcon size={16} />
                    Upload Updated W8/W9 PDF
                  </>
                ) : (
                  <>
                    <UploadIcon size={16} />
                    Choose PDF File
                  </>
                )}
              </div>
              <input
                onChange={handleFileChange}
                disabled={loading || noEmail}
                type='file'
                className='absolute invisible w-full top-0 bottom-0 z-10'
                id='file input'
              />
            </label>
            <div className='border border-web-gray w-full flex items-center font-semibold h-8 px-2 rounded-sm'>
              {file?.name}
            </div>
            <button
              onClick={openModal}
              disabled={!file}
              className={`flex gap-2 h-8 items-center justify-center text-center
                ${file ? 'btn-requirements cursor-pointer' : 'btn-default cursor-not-allowed'}
              `}
            >
              {'Send'}
              {loading && <LoadingIcon />}
            </button>
            {showTaxModal && (
              <ModalDefault setShowModal={setShowTaxModal} title='Submit tax form' footerRight={submitBtn}>
                <div className='flex gap-2'>
                  <input
                    className='checkbox'
                    checked={compliance.signed}
                    id='signed'
                    type='checkbox'
                    onClick={updateCompliance}
                  />
                  <label htmlFor='signed'>I have properly signed my tax documents</label>
                </div>{' '}
                <div className='flex gap-2'>
                  <input
                    className='checkbox'
                    checked={compliance.dated}
                    id='dated'
                    type='checkbox'
                    onClick={updateCompliance}
                  />

                  <label htmlFor='dated'>I have properly dated my tax documents</label>
                </div>
                <div className='flex gap-2'>
                  <input
                    className='checkbox'
                    checked={compliance.relevant}
                    id='relevant'
                    type='checkbox'
                    onClick={updateCompliance}
                  />

                  <label htmlFor='relevant'>I have completed all relevant sections of my tax documents</label>
                </div>
              </ModalDefault>
            )}
          </form>
          <div className=''>
            *W-8 forms are{' '}
            <Link
              href={'https://www.irs.gov/'}
              rel='noopener norefferer'
              target='_blank'
              className='text-link-colour hover:underline col-span-2'
            >
              Internal Revenue Service
            </Link>{' '}
            (IRS) forms that foreign individuals and businesses must file to verify their country of residence for tax
            purposes, certifying that they qualify for a lower rate of tax withholding.
          </div>
        </>
      )}
      {w8formResponse && (
        <div className='bg-info border-info-strong border p-4 rounded-sm'>
          <MessageHTML />
        </div>
      )}
      {showModal && <W8FormModal setShowModal={setShowModal} />}
    </section>
  );
};

export default W8Requirement;
