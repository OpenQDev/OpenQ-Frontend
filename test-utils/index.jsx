// Third party
import React from 'react';
// Cusotm
import { render, renderHook } from '@testing-library/react';
import { cleanup } from '@testing-library/react';

import StoreProvider from '../store/Store/TestStoreProvider';
import AuthProvider from '../store/AuthStore/TestAuthProvider';
import FundProvider from '../components/FundBounty/FundStore/TestFundProvider';
import TokenProvider from '../components/TokenSelection/TokenStore/TokenProvider';

// Add in any providers here if necessary:
// test-utils.js
// (ReduxProvider, ThemeProvider, etc)
afterEach(() => {
  cleanup();
});

const customRender = (ui, options = {}, storeProps, authProps) => {
  const IntersectionObserverMock = vi.fn(() => ({
    disconnect: vi.fn(),
    observe: vi.fn(),
    takeRecords: vi.fn(),
    unobserve: vi.fn(),
  }));
  vi.stubGlobal('IntersectionObserver', IntersectionObserverMock);
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <TokenProvider>{children}</TokenProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return render(ui, { wrapper: Providers, ...options });
};
const customHookRender = (hook, options = {}, storeProps, authProps, fundProps) => {
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <FundProvider StoreProps={fundProps}>
            <TokenProvider>{children}</TokenProvider>
          </FundProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return renderHook(hook, { wrapper: Providers, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, customHookRender as hookRender };
