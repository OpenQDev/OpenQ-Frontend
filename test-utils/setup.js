import { expect, beforeEach } from 'vitest';
import { cleanup } from '@testing-library/react';
import matchers from '@testing-library/jest-dom/matchers';

expect.extend(matchers);
beforeEach(() => {
  cleanup();
});
process.on('unhandledRejection', (reason) => {
  // eslint-disable-next-line no-console
  console.log(`FAILED TO HANDLE PROMISE REJECTION`);
  throw reason;
});
