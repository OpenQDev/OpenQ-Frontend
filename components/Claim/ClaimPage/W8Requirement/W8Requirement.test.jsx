import { render } from '../../../../test-utils';
import React from 'react';
import W8Requirement from '.';
import Constants from '../../../../test-utils/constant';

describe('W8Requirement', () => {
  it('should match initial DOM Snapshot', () => {
    const { asFragment } = render(<W8Requirement bounty={Constants.bounty3} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
