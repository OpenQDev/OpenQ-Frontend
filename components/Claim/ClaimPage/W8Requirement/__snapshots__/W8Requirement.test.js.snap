// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`W8Requirement should match initial DOM Snapshot 1`] = `
<DocumentFragment>
  <section
    class="flex flex-col gap-3"
  >
    <h4
      class="text-2xl flex content-center items-center gap-2 border-b border-gray-700 pb-2"
    >
      Form W8/W9*
      <div
        class="bg-info border-2 border-info-strong text-sm px-2 border rounded-full h-6"
      >
        Required
      </div>
    </h4>
    <div>
      <div>
        Please complete and upload a W-8/W-9 form. Choose one of five types, depending on your entity. We encourage you to consult with you own tax or financial adviser to determine which form is appropriate for you or ask in our
        <div>
          <a
            class="text-link-colour hover:underline col-span-2"
            href="https://discord.gg/puQVqEvVXn"
            rel="noopener norefferer"
            target="_blank"
          >
            discord
          </a>
           for help.
        </div>
      </div>
      <p>
        Please add an email to your profile so that we can send you a copy of the your submitted form.
      </p>
      <div
        class="false py text-lg"
      >
        <form
          class="font-normal max-w-[500px] gap-4"
        >
          <div>
            <div
              class=" text-sm justify-between my-4"
            >
              <label
                class="block font-semibold mb-1.5"
                for="invoicingEmail"
              >
                Email 
              </label>
              <input
                class="input-field w-full max-w-[400px] bg-input-bg px-3 py-[5px] border-border-muted invalid:border-danger"
                id="invoicingEmail"
                required=""
                type="text"
                value=""
              />
            </div>
            <div
              class="note my-4"
            >
              This email will be used by organizations to communicate with you. OpenQ will also send copies of any documents you send to organizers via our interface.
            </div>
          </div>
          <input
            class="btn-primary bg-green  text-sm py-1 px-2 text-center w-20 cursor-pointer"
            type="submit"
            value="Update"
          />
        </form>
      </div>
    </div>
    <div>
      Explore our W8/W9 templates 
      <button
        class="text-blue-500 hover:underline"
      >
        here
      </button>
      .
    </div>
    <div
      class="font-semibold flex gap-2 group w-fit"
    >
      Upload
    </div>
    <form
      class="flex gap-2 items-center flex-wrap md:flex-nowrap"
    >
      <div
        class="relative flex w-fit whitespace-nowrap gap-2 h-8 items-center justify-center text-center btn-default-disabled cursor-not-allowed"
      >
        Pending
      </div>
      <label
        class="relative flex items-center"
        for="file input"
      >
        <div
          class="relative flex w-fit whitespace-nowrap gap-2 h-8 items-center justify-center text-center cursor-not-allowed btn-primary"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-upload"
            fill="currentColor"
            focusable="false"
            height="16"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M2.75 14A1.75 1.75 0 0 1 1 12.25v-2.5a.75.75 0 0 1 1.5 0v2.5c0 .138.112.25.25.25h10.5a.25.25 0 0 0 .25-.25v-2.5a.75.75 0 0 1 1.5 0v2.5A1.75 1.75 0 0 1 13.25 14Z"
            />
            <path
              d="M11.78 4.72a.749.749 0 1 1-1.06 1.06L8.75 3.811V9.5a.75.75 0 0 1-1.5 0V3.811L5.28 5.78a.749.749 0 1 1-1.06-1.06l3.25-3.25a.749.749 0 0 1 1.06 0l3.25 3.25Z"
            />
          </svg>
          Upload Updated W8/W9
        </div>
        <input
          class="absolute invisible w-full top-0 bottom-0 z-10"
          disabled=""
          id="file input"
          type="file"
        />
      </label>
      <div
        class="border border-web-gray w-full flex items-center font-semibold h-8 px-2 rounded-sm"
      />
      <button
        class="flex gap-2 h-8 items-center justify-center text-center
                btn-default cursor-not-allowed
              "
        disabled=""
      >
        Sent
      </button>
    </form>
    <div
      class=""
    >
      *W-8 forms are 
      <a
        class="text-link-colour hover:underline col-span-2"
        href="https://www.irs.gov/"
        rel="noopener norefferer"
        target="_blank"
      >
        Internal Revenue Service
      </a>
       (IRS) forms that foreign individuals and businesses must file to verify their country of residence for tax purposes, certifying that they qualify for a lower rate of tax withholding.
    </div>
  </section>
</DocumentFragment>
`;
